namesys/routing.go

```diff
@@ -65,24 +66,38 @@ func (r *routingResolver) Resolve(ctx context.Context, name string) (u.Key, erro
 
 	// name should be a public key retrievable from ipfs
 	// /ipfs/<name>
-	key := u.Key("/pk/" + string(hash))
-	pkval, err := r.routing.GetValue(ctx, key)
-	if err != nil {
-		log.Warning("RoutingResolve PubKey Get failed.")
-		return "", err
-	}
+	var pubkey ci.PubKey
+	if dht, ok := r.routing.(routing.PubKeyFetcher); ok {
+		// If we have a DHT as our routing system, use optimized fetcher
+		pk, err := dht.GetPublicKey(ctx, peer.ID(hash))
+		if err != nil {
+			log.Warning("RoutingResolve PubKey Get failed.")
+			return "", err
+		}
+		pubkey = pk
+	} else {
+		key := u.Key("/pk/" + string(hash))
+		pkval, err := r.routing.GetValue(ctx, key)
+		if err != nil {
+			log.Warning("RoutingResolve PubKey Get failed.")
+			return "", err
+		}
 
-	// get PublicKey from node.Data
-	pk, err := ci.UnmarshalPublicKey(pkval)
-	if err != nil {
-		return "", err
+		// get PublicKey from node.Data
+		pk, err := ci.UnmarshalPublicKey(pkval)
+		if err != nil {
+			return "", err
+		}
+
+		pubkey = pk
```

Comment:
this is a pretty big function-- maybe abstract this piece (grabbing the key from routing) into a 

```
func (r *routingResolver) getPublicKey(...) {}
```